### PR TITLE
Add read from relay feature

### PR DESCRIPTION
**Pull request by kriptonix**
This code adds basic functionality to read events from Nostr relay. Send request to relay, get response and close WebSocket connection. For continuous WebSocket connection code needs to be additionally upgraded.

Create new subscription and set subscription ID. I use simple code to generate random 64 character ID for subscription:
```
$subscription = new Subscription();
$subscriptionId = $subscription->setId();
```
Define as many filters as you like:
```
$filter1 = new Filter();
$filter1->setAuthors(<array of pubkeys>);
$filter1->setKinds(<array of kinds>);
$filter1->setLimit(<integer>);
$filter1->setSince(<timestamp integer>);
$filter1->setUntil(<timestamp integer>);
$filter1->setETag(<array of #e tags>);
$filter1->setPTag(<array of #p tags>]);
```
```
$filter2 = new Filter();
$filter2->setKinds([1]);
$filter2->setLimit(3);
```
Combine all filters into one array:
`$filters = [$filter1, $filter2];`

Create request message:
`$requestMessage = new RequestMessage($subscriptionId, $filters);`

Create request and send it to relay:
```
$request = new Request($relayUrl, $requestMessage);
$result = $request->send();
```

### CODE REVIEW

**Review by Sebastix — CHANGES_REQUESTED** [review]
Thanks for this PR!
My review on this is work in progress, but here are my first comments @kriptonix

File: src/Request/Request.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
+<?php
+
+declare(strict_types=1);
+
+namespace swentel\nostr\Request;
+
+use swentel\nostr\Message\RequestMessage;
```
Comment (Sebastix) [inline src/Request/Request.php:7]: This is an unused statement which can be removed

File: src/Filter/Filter.php [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+    /**
+     * A list of #e tag values (list of event ids)
+     */
+    public array $etags;
+
+    /**
+     * A list of #p tag values (list of pubkeys).
+     */
+    public array $ptags;
```
Comment (Sebastix) [inline src/Filter/Filter.php:34]: FYI Nostr event can have lowercase and uppercase p tags (zap receipt kind `9735` event for example)

File: src/Filter/Filter.php [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+     * Set the #p tag for the Filter object.
+     *
+     * @param array $ptag The array of tag to set.
+     */
+    public function setPTag(array $ptags): static
+    {
+        // Check IF array contain exact 64-character lowercase hex values
+        foreach($ptags as $tag){
+            if(!$this->isLowercaseHex($tag)){
+                throw new \RuntimeException("#e tags must be an array of 64-character lowercase hex values");
```
Comment (Sebastix) [inline src/Filter/Filter.php:104]: Small typo here as it mentions #e tags instead of #p tags

File: src/Filter/Filter.php [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        $this->etags = $etags;
+        return $this;
+    }
+
+    /**
+     * Set the #p tag for the Filter object.
+     *
+     * @param array $ptag The array of tag to set.
+     */
+    public function setPTag(array $ptags): static
```
Comment (Sebastix) [inline src/Filter/Filter.php:99]: I would propose to change the function name to setLowercasePTags as there also can be a setUppercasePTags

**Review by Sebastix — CHANGES_REQUESTED** [review]

File: src/Filter/Filter.php (hunk 4 above)
Comment (kriptonix) [reply src/Filter/Filter.php:99]: Good point. I also used isLowercaseHex() for setting #e tags so might be useful to also change setETag() to setLowercaseETags

**Pull request merged**